gcloud-java-core/src/main/java/com/google/gcloud/BaseIamPolicy.java

```diff
@@ -0,0 +1,297 @@
+     * Adds one or more bindings to the policy.
+     */
+    public B addBinding(R role, List<Identity> identities) {
+      bindings.put(role, ImmutableList.copyOf(identities));
+      return self();
+    }
+
+    /**
+     * Removes the specified ACL.
+     */
+    public B removeBinding(R role) {
+      bindings.remove(role);
+      return self();
+    }
+
+    /**
+     * Sets the policy's etag.
+     *
+     * <p>Etags are used for optimistic concurrency control as a way to help prevent simultaneous
+     * updates of a policy from overwriting each other. It is strongly suggested that systems make
+     * use of the etag in the read-modify-write cycle to perform policy updates in order to avoid
+     * race conditions. An etag is returned in the response to getIamPolicy, and systems are
+     * expected to put that etag in the request to setIamPolicy to ensure that their change will be
+     * applied to the same version of the policy.  If no etag is provided in the call to
+     * setIamPolicy, then the existing policy is overwritten blindly.
+     */
+    protected B etag(String etag) {
+      this.etag = etag;
+      return self();
+    }
+
+    /**
+     * Sets the version of the policy. The default version is 0, meaning roles that are in alpha
+     * (non-legacy) roles are not permitted. If the version is 1, you may use roles other than
+     * "owner", "editor", and "viewer".
+     */
+    protected B version(int version) {
+      this.version = version;
+      return self();
+    }
+
+    @SuppressWarnings("unchecked")
+    private B self() {
+      return (B) this;
+    }
+
+    public abstract BaseIamPolicy<R> build();
+  }
+
+  protected BaseIamPolicy(BaseBuilder<R, ? extends BaseBuilder<R, ?>> builder) {
+    this.bindings = builder.bindings;
+    this.etag = builder.etag;
+    this.version = builder.version;
+  }
+
+  /**
+   * The list of ACLs specified in the policy.
+   */
+  public Map<R, List<Identity>> bindings() {
+    return bindings;
+  }
+
+  /**
+   * The policy's etag.
+   *
+   * <p>Etags are used for optimistic concurrency control as a way to help prevent simultaneous
+   * updates of a policy from overwriting each other. It is strongly suggested that systems make
+   * use of the etag in the read-modify-write cycle to perform policy updates in order to avoid
+   * race conditions. An etag is returned in the response to getIamPolicy, and systems are
+   * expected to put that etag in the request to setIamPolicy to ensure that their change will be
+   * applied to the same version of the policy.  If no etag is provided in the call to
+   * setIamPolicy, then the existing policy is overwritten blindly.
+   */
+  public String etag() {
+    return etag;
+  }
+
+  /**
+   * The version of the policy. The default version is 0.
+   */
+  public int version() {
+    return version;
+  }
+
+  public int baseHashCode() {
+    return Objects.hash(bindings, etag, version);
+  }
+
+  public boolean baseEquals(Object obj) {
```

Comment:
Alternative to protected is to make rename them to hashCode and equals, make them public and final and change equals to check for class equality rather than instanceof.